vyper/semantics/analysis/imports.py

```diff
@@ -0,0 +1,328 @@
+    # track of imports in the current module
+    _imports: list[dict] = field(default_factory=list)
+
+    @property
+    def imported_modules(self):
+        return self._imports[-1]
+
+    @property
+    def current_module(self):
+        return self._path[-1]
+
+    def push_path(self, module_ast: vy_ast.Module) -> None:
+        if module_ast in self._path:
+            cycle = self._path + [module_ast]
+            raise ImportCycle(" imports ".join(f'"{t.path}"' for t in cycle))
+
+        self._path.append(module_ast)
+        self._imports.append({})
+
+    def pop_path(self, expected: vy_ast.Module) -> None:
+        popped = self._path.pop()
+        if expected != popped:
+            raise CompilerPanic("unreachable")
+        self._imports.pop()
+
+    @contextlib.contextmanager
+    def enter_path(self, module_ast: vy_ast.Module) -> Iterator[None]:
+        self.push_path(module_ast)
+        try:
+            yield
+        finally:
+            self.pop_path(module_ast)
+
+
+class ImportAnalyzer:
+    def __init__(self, input_bundle: InputBundle, graph: _ImportGraph):
+        self.input_bundle = input_bundle
+        self.graph = graph
+        self._ast_of: dict[int, vy_ast.Module] = {}
+
+        self.seen: set[int] = set()
+
+        self.integrity_sum = None
+
+    def resolve_imports(self, module_ast: vy_ast.Module):
+        self._resolve_imports_r(module_ast)
+        self.integrity_sum = self._calculate_integrity_sum_r(module_ast)
+
+    def _calculate_integrity_sum_r(self, module_ast: vy_ast.Module):
+        acc = [sha256sum(module_ast.full_source_code)]
+        for s in module_ast.get_children((vy_ast.Import, vy_ast.ImportFrom)):
+            info = s._metadata["import_info"]
+
+            if info.compiler_input.path.suffix in (".vyi", ".json"):
+                # NOTE: this needs to be redone if interfaces can import other interfaces
+                acc.append(info.compiler_input.sha256sum)
+            else:
+                acc.append(self._calculate_integrity_sum_r(info.parsed))
+
+        return sha256sum("".join(acc))
+
+    def _resolve_imports_r(self, module_ast: vy_ast.Module):
+        if id(module_ast) in self.seen:
+            return
+        with self.graph.enter_path(module_ast):
+            for node in module_ast.body:
+                if isinstance(node, vy_ast.Import):
+                    self._handle_Import(node)
+                elif isinstance(node, vy_ast.ImportFrom):
+                    self._handle_ImportFrom(node)
+        self.seen.add(id(module_ast))
+
+    def _handle_Import(self, node: vy_ast.Import):
+        # import x.y[name] as y[alias]
+
+        alias = node.alias
+
+        if alias is None:
+            alias = node.name
+
+        # don't handle things like `import x.y`
+        if "." in alias:
+            msg = "import requires an accompanying `as` statement"
+            suggested_alias = node.name[node.name.rfind(".") :]
```

Comment:
this will include the '.', no?

Reply:
```
  contract "tests/custom/test4.vy:1", line 1:0 
  ---> 1 import lib1.foo
  -------^
       2 import lib2

  (hint: try `import lib1.foo as .foo`)

```

Reply:
nice find. i want to keep this PR as semantically neutral as possible (fewest semantic changes), maybe we should kick it to a new issue?

Reply:
sure, created a new issue: https://github.com/vyperlang/vyper/issues/4278